Implement GroupCast Test Cases

stefan-silabs opened the pull request:
#### Summary

Test script implementation for TC_GCAST_2_1, TC_GCAST_2_3, TC_GCAST_2_4, TC_GCAST_2_5, TC_GCAST_2_6 and TC_GCAST_2_7.

#### Related issues

Tasks:
https://github.com/project-chip/connectedhomeip/issues/41852
https://github.com/project-chip/connectedhomeip/issues/41854
https://github.com/project-chip/connectedhomeip/issues/41855
https://github.com/project-chip/connectedhomeip/issues/41856
https://github.com/project-chip/connectedhomeip/issues/41857
https://github.com/project-chip/connectedhomeip/issues/41859

#### Testing

Currently cannot test.

#### Readability checklist

The checklist below will help the reviewer finish PR review in time and keep the
code readable:

-   [ ] PR title is
        [descriptive](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html#title-formatting)
-   [ ] Apply the
        [_“When in Rome…”_](https://project-chip.github.io/connectedhomeip-doc/style/CODING_STYLE_GUIDE.html)
        rule (coding style)
-   [ ] PR size is short
-   [ ] Try to avoid "squashing" and "force-update" in commit history
-   [ ] CI time didn't increase

See: [Pull Request Guidelines](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html)

stefan-silabs committed:
begin implementation gcast 2.1, 2.2, 2.3

stefan-silabs committed:
draft implementation of ot TC_GCAST_2_1 and TC_GCAST_2_3

stefan-silabs committed:
fix typo in retrieve_join_group_cmd_endpoints

stefan-silabs committed:
add GCAST_2_4 and GCAST_2_5 implementation, update GCAST_2_3

stefan-silabs committed:
add GCAST_2_6 implementation, update get endpoints list in other GCAST test cases

stefan-silabs committed:
add GCAST_2_6, missed in previous commit

stefan-silabs committed:
add GCAST_2_7 and update other GCAST test cases to use secrets lib for keys

stefan-silabs committed:
Merge branch 'master' of https://github.com/project-chip/connectedhomeip into task/implement_groupcast_script

stefan-silabs committed:
minor fixes pr copilot comments

stefan-silabs committed:
add missing step for TC_GCAST_2_5 and fix TC_GCAST_2_1 syntax error

stefan-silabs committed:
fix GCAST_2_6 and GCAST_2_7 imports and test names

stefan-silabs committed:
address most comments by Junior and copilot

tcarmelveilleux commented on src/python_testing/TC_GCAST_2_1.py:
```diff
@@ -0,0 +1,78 @@
+
+        self.step(3)
+        M_max = await self.read_single_attribute_check_success(groupcast_cluster, max_membership_count_attribute)
+        asserts.assert_true(M_max >= 10, "MaxMembershipCount attribute should be >= 10")
```
assert_greater_equal.

Please replace *every* `assert_true(X <OPERATOR> Y, "xxx")` with the correct *specific* assertion.

tcarmelveilleux commented on src/python_testing/TC_GCAST_2_4.py:
```diff
@@ -0,0 +1,207 @@
+
+    async def get_feature_map(self):
+        """Get supported features."""
+        feature_map = await self.read_single_attribute_check_success(
```
This must force endpoint 0

stefan-silabs commented on src/python_testing/TC_GCAST_2_3.py:
```diff
@@ -0,0 +1,286 @@
+                                 f"Send UpdateGroupKey command error should be {Status.AlreadyExists} instead of {e.status}")
+
+        self.step(5)
+        time.sleep(gracePeriod * 1.1)
```
will change.

stefan-silabs commented on src/python_testing/TC_GCAST_2_3.py:
```diff
@@ -0,0 +1,286 @@
+
+        self.step(2)
+        keyID3 = 3
+        gracePeriod = 5
```
will change

stefan-silabs commented on src/python_testing/TC_GCAST_2_3.py:
```diff
@@ -0,0 +1,286 @@
+            TestStep(12, "Update Group G1 with KeyID already used by another group. "),
+        ]
+
+    def pics_TC_GCAST_2_3(self) -> list[str]:
```
will adjust for this

stefan-silabs committed:
address comments by tcarmelveilleux

stefan-silabs committed:
fix test name in GCAST_2_6

tcarmelveilleux commented on src/python_testing/TC_GCAST_2_1.py:
```diff
@@ -68,11 +76,20 @@ async def test_TC_GCAST_2_1(self):
         membership = await self.read_single_attribute_check_success(groupcast_cluster, membership_attribute)
-        asserts.assert_is_instance(membership, list, "Membership attribute should be a list")
+        asserts.assert_is_instance(membership, list, "Membership attribute list should be a list")
+        asserts.assert_equal(membership, [], "Membership attribute list should be empt.y")
```
```suggestion
        asserts.assert_equal(membership, [], "Membership attribute list should be empty.")
```

tcarmelveilleux commented on src/python_testing/TC_GCAST_2_1.py:
```diff
@@ -38,10 +38,18 @@
+    Clusters.ColorControl.id,
+    Clusters.SceneManagement.id
+}
+
```
Perhaps move this right in the method for now...

tcarmelveilleux commented on src/python_testing/TC_GCAST_2_3.py:
```diff
@@ -40,12 +40,92 @@
+from matter.testing.matter_testing import AttributeMatcher, MatterBaseTest, TestStep, async_test_body, default_matter_test_main, has_cluster, run_if_endpoint_matches
 from matter.interaction_model import InteractionModelError, Status
 
+from src.python_testing.TC_GCAST_2_1 import is_groupcast_supporting_cluster
```
This is odd. I did not expect this type of import. Suggest making a TC_GCAST_common.py file for these utilities.

tcarmelveilleux commented on src/python_testing/TC_GCAST_2_3.py:
```diff
@@ -40,12 +40,92 @@
 
 
+def membership_entry_matcher(
+        group_id: int, key_id: int = None,
```
One arg per line please

tcarmelveilleux commented on src/python_testing/TC_GCAST_2_3.py:
```diff
@@ -40,12 +40,92 @@
+        group_id: int, key_id: int = None,
+        expiring_key_id: int = None,
+        has_auxiliary_acl: str = None,
+        endpoints: list = None,
```
```suggestion
        expiring_key_id: Optional[int] = None,
        has_auxiliary_acl: Optional[str] = None,
        endpoints: Optional[list] = None,
```

tcarmelveilleux commented on src/python_testing/TC_GCAST_2_3.py:
```diff
@@ -263,12 +326,12 @@ async def test_TC_GCAST_2_3(self):
 
         self.step(11)
-        gracePeriodInvalidLimit = 86401
+        gracePeriodSecondsInvalidLimit = 86401
```
```suggestion
        gracePeriodSecondsInvalidLimit = 86400 + 1
```

tcarmelveilleux commented on src/python_testing/TC_GCAST_2_6.py:
```diff
@@ -66,29 +69,33 @@ def steps_TC_G_2_6(self):
         """
-        Get the JoinGroup cmd endpoints list when the listener feature is enabled. Represent the 1 first valid non-root and
+        Get the JoinGroup cmd endpoints list when the listener feature is enabled. Represents the 1 first valid non-root and
         non-aggregator endpoint.
```
```suggestion
        non-aggregator endpoint with a groupcast-supporting cluster.
```

stefan-silabs committed:
move common logic to TC_GCAST_common, update test cases and update based on comments

PriyaMalhotra3 committed:
WIP boilerplate for TC_GCAST_2_2

PriyaMalhotra3 committed:
Finished first pass of TC_GCAST_2_2; will come back to clean up comments

PriyaMalhotra3 committed:
fixed catching exceptions

PriyaMalhotra3 committed:
Adding step description above each step as a comment to help reader follow tests in-line.

PriyaMalhotra3 committed:
Addressing rest of comments from @tcarmelveilleux

tcarmelveilleux committed:
Revert commits by @PriyaMalhotra3 on wrong branch

tcarmelveilleux committed:
Merge remote-tracking branch 'upstream/master' into task/implement_groupcast_script

tcarmelveilleux committed:
Fix linter

- Fix copyright dates
- Fix the following lints:

```
Error: src/python_testing/TC_GCAST_2_1.py:40:38: F401 `matter.interaction_model.Status` imported but unused
Error: src/python_testing/TC_GCAST_2_1.py:41:69: F401 `matter.testing.matter_testing.async_test_body` imported but unused
Error: src/python_testing/TC_GCAST_2_1.py:57:16: RET504 Unnecessary assignment to `pics` before `return` statement
Error: src/python_testing/TC_GCAST_2_3.py:43:43: F401 `matter.testing.matter_testing.AttributeMatcher` imported but unused
Error: src/python_testing/TC_GCAST_2_3.py:43:87: F401 `matter.testing.matter_testing.async_test_body` imported but unused
Error: src/python_testing/TC_GCAST_2_3.py:77:16: RET504 Unnecessary assignment to `pics` before `return` statement
Error: src/python_testing/TC_GCAST_2_3.py:154:9: ASYNC251 Async functions should not call `time.sleep`
Error: src/python_testing/TC_GCAST_2_4.py:43:69: F401 `matter.testing.matter_testing.async_test_body` imported but unused
Error: src/python_testing/TC_GCAST_2_4.py:71:16: RET504 Unnecessary assignment to `pics` before `return` statement
Error: src/python_testing/TC_GCAST_2_5.py:43:69: F401 `matter.testing.matter_testing.async_test_body` imported but unused
Error: src/python_testing/TC_GCAST_2_5.py:80:16: RET504 Unnecessary assignment to `pics` before `return` statement
Error: src/python_testing/TC_GCAST_2_6.py:43:69: F401 `matter.testing.matter_testing.async_test_body` imported but unused
Error: src/python_testing/TC_GCAST_2_6.py:72:16: RET504 Unnecessary assignment to `pics` before `return` statement
Error: src/python_testing/TC_GCAST_2_7.py:41:69: F401 `matter.testing.matter_testing.async_test_body` imported but unused
Error: src/python_testing/TC_GCAST_2_7.py:66:16: RET504 Unnecessary assignment to `pics` before `return` statement
Error: src/python_testing/TC_GCAST_common.py:53:17: RET507 Unnecessary `else` after `continue` statement
Error: src/python_testing/TC_GCAST_common.py:152:86: W292 No newline at end of file
```

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

andy31415 commented on src/python_testing/TC_GCAST_2_7.py:
Add this file to https://github.com/project-chip/connectedhomeip/blob/master/src/python_testing/test_metadata.yaml#L3 since it is a common file and does not require running as a test.

tcarmelveilleux commented on src/python_testing/TC_GCAST_2_1.py:
```diff
@@ -0,0 +1,78 @@
+    @async_test_body
+    async def test_TC_GCAST_2_1(self):
+        if self.matter_test_config.endpoint is None:
+            self.matter_test_config.endpoint = 1
```
Please do remove all the self.matter_test_config.endpoint setting everywhere.

stefan-silabs replied:
will remove

stefan-silabs committed:
Merge commit '383d4d4113857a601912efcff32cd96dd66087b8' into task/implement_groupcast_script

stefan-silabs committed:
fix imports, resolve comments

stefan-silabs committed:
fix ci arguments block, fix typo in GCAST_common

stefan-silabs committed:
fix TC_GCAST_common imports in TC_GCAST test cases, and add test cases to test_metadata.yaml to prevent from running in CI for now

stefan-silabs committed:
update reason for TC_GCAST in test_metatadata.yaml, remove set endpoint blocks in TC_GCAST test cases

restyled-commits committed:
Restyled by prettier-yaml

restyled-commits committed:
Restyled by isort
